### PR TITLE
Sherin/timeseries upgrade

### PR DESCRIPTION
**Pull request by ann-sherin**
This adds capability to perform time-series upgrade analysis in DISCO

### CODE REVIEW

File: .github/workflows/pull_request_tests.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
```
Comment (daniel-thom) [inline .github/workflows/pull_request_tests.yml:12]: Why this change?

File: disco/cli/upgrade_cost_analysis.py [hunk 2]
```diff
@@ -299,18 +299,22 @@ def run_job(job, config, jobs_output_dir, file_log_level):
         global_config["upgrade_simulation_params"]["pydss_controller"] = (
             config.pydss_controllers.pv_controller.dict(),
         )
+    # create a simulation params dictionary
+    rem_list = ["thermal_upgrade_params", "voltage_upgrade_params", "jobs", "enable_pydss_controllers", "pydss_controllers"]
+    simulation_params_config = dict(config)
```
Comment (daniel-thom) [inline disco/cli/upgrade_cost_analysis.py:304]: I'm surprised this works. I would have used `config.dict()`. Is the result the same?

File: disco/cli/upgrade_cost_analysis.py [hunk 3]
```diff
@@ -299,18 +299,22 @@ def run_job(job, config, jobs_output_dir, file_log_level):
...
         job_global_config=global_config,
         output=jobs_output_dir,
     )
     simulation.run(
-        dc_ac_ratio=global_config["dc_ac_ratio"],
         enable_pydss_solve=global_config["upgrade_simulation_params"]["enable_pydss_controller"],
         pydss_controller_model=config.pydss_controllers.pv_controller,
         thermal_config=global_config["thermal_upgrade_params"],
         voltage_config=global_config["voltage_upgrade_params"],
+        simulation_params_config=simulation_params_config,
```
Comment (daniel-thom) [inline disco/cli/upgrade_cost_analysis.py:317]: Doesn't this include all of the other parameters? Maybe you did this for convenience, but should we consider a larger change?

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 4]
```diff
@@ -72,6 +72,9 @@ def reload_dss_circuit(dss_file_list, commands_list=None,  **kwargs):
             check_dss_run_command(command_string)
             if "new " in command_string.lower():
                 check_dss_run_command("CalcVoltageBases")
+    solve = kwargs.get("solve", True)
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:75]: I'll suggest that we start defining these parameters in the function.
```
reload_dss_circuit(dss_file_list, commands_list=None, solve=True)
```

File: .github/workflows/pull_request_tests.yml (hunk 1 above)
Comment (ann-sherin) [reply .github/workflows/pull_request_tests.yml:12]: Uh, very weird. I didn't really change this file. But maybe during some merge this must've happened. 
I'll revert this change to os: [ubuntu-latest, macos-latest, windows-latest]

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 5]
```diff
@@ -1128,7 +1138,69 @@ def get_thermal_equipment_info(compute_loading, equipment_type, upper_limit=None
...
+        elif equipment_type.lower() == "voltage":
+            bus_voltages_df, undervoltage_bus_list, overvoltage_bus_list, buses_with_violations = get_bus_voltages_instance(
+                    voltage_upper_limit=voltage_upper_limit, voltage_lower_limit=voltage_lower_limit, raise_exception=raise_exception, **kwargs)
+            bus_voltages_df.set_index("name", inplace=True)
+            comparison_dict[key_name] = bus_voltages_df
     
+    # compare all dataframe, and create one that contains all worst loading conditions (across all multiplier conditions)
+    if (equipment_type.lower() == "transformer") or (equipment_type.lower() == "line"):
+        loading_df = compare_multiple_dataframes(comparison_dict, deciding_column_name, comparison_type="max")
+        return loading_df
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1193]: Can you make it so that this function always returns the same type? It's possible that you could make different functions for transformers/lines vs voltage. That would eliminate a lot of your conditional statements.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 6]
```diff
@@ -1128,7 +1138,69 @@ def get_thermal_equipment_info(compute_loading, equipment_type, upper_limit=None
...
+    elif equipment_type == "voltage":  # pop out of dictionary since kwargs are also passed to voltage function
+        voltage_upper_limit = kwargs.pop("voltage_upper_limit")
+        voltage_lower_limit = kwargs.pop("voltage_lower_limit")
+        raise_exception = kwargs.pop("raise_exception")
+        deciding_column_dict = kwargs.pop("deciding_column_dict")
+    else:
+        raise Exception(f"Incorrect equipment type {equipment_type} provided. Possible values: line, transformer, voltage.")
+        
+    comparison_dict = {}
+    sol = dss.Solution  # opendss solution interface
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1161]: I'll suggest that it's more clear to use `dss.Solution` throughout the code.

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 7]
```diff
@@ -1128,7 +1138,69 @@ def get_thermal_equipment_info(compute_loading, equipment_type, upper_limit=None
...
+    dss.run_command("Set mode=yearly number=1")  # 1 power flow per solve and I guess stepsize remains the same
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1162]: `dss.Text.Command` raises an exception on failure, right? Want to use it everywhere?

Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1162]: "I guess" is not a great comment in code. The response from the next developer will be, "so you're making me look this up to be confident that it's true?"

File: disco/extensions/upgrade_simulation/upgrades/common_functions.py [hunk 8]
```diff
@@ -1128,7 +1138,69 @@ def get_thermal_equipment_info(compute_loading, equipment_type, upper_limit=None
...
+    for present_step in range(startH, num_pts, time_step):
+        key_name = "timepoint_"+str(present_step)
+        logger.debug(f"Present step: {present_step}")
+        dss.run_command("Set Controlmode=Static") 
+        dss.Text.Command(f"set hour = {present_step}")
+        sol.Solve()  # solves for specific hour that has been set
+        logger.debug(dss.LoadShape.Name())
+        logger.debug(dss.LoadShape.PMult()[present_step])
+        logger.debug(dss.Loads.Name())
+        logger.debug(sum(dss.CktElement.Powers()[::2]))
```
Comment (daniel-thom) [inline disco/extensions/upgrade_simulation/upgrades/common_functions.py:1176]: One log statement with each name-value pair?

File: tests/integration/test_generic_upgrade.py [hunk 9]
```diff
@@ -40,18 +40,18 @@ def test_upgrades(cleanup):
...
-    run_cmd = f"disco upgrade-cost-analysis run {TEST_UPGRADES_CONFIG_FILE_2} -o {OUTPUT}"
-    assert run_command(run_cmd) == 0
-    verify_upgrade_results(OUTPUT)
-    run_cmd = f"disco upgrade-cost-analysis run {TEST_UPGRADES_CONFIG_FILE_3} -o {OUTPUT}"
-    assert run_command(run_cmd) == 0
-    verify_upgrade_results(OUTPUT)
-    run_cmd = f"disco upgrade-cost-analysis run {TEST_UPGRADES_CONFIG_FILE_4} -o {OUTPUT}"
-    assert run_command(run_cmd) == 0
-    verify_upgrade_results(OUTPUT)
+    #run_cmd = f"disco upgrade-cost-analysis run {TEST_UPGRADES_CONFIG_FILE_1} -o {OUTPUT}"
```
Comment (daniel-thom) [inline tests/integration/test_generic_upgrade.py:43]: Why is this commented out?
